config/config/src/main/java/io/helidon/config/EnumMapperProvider.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.config;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import io.helidon.config.spi.ConfigMapperProvider;
+
+/**
+ * Built-in mapper for {@code enum}s.
+ *
+ * <p>
+ *     This mapper attempts to match strings in the config source to enum values as follows:
+ *     <ul>
+ *         <li>The mapper treats hyphens ('-') in config strings as underscores when comparing to enum value names.</li>
+ *         <li>If the matcher finds a <em>case-sensitive</em> match with an enum value name, then that enum value matches.</li>
+ *         <li>If the matcher finds exactly one <em>case-insensitive</em> match, that enum value matches.</li>
+ *         <li>If the matcher finds no matches or multiple matches, throw a
+ *         {@link io.helidon.config.ConfigEnumMappingException} with a message explaining the problem.</li>
+ *     </ul>
+ *     These conversions are intended to maximize ease-of-use for authors of config sources so the values need not be
+ *     upper-cased nor punctuated with underscores rather than the more conventional (in config at least) hyphen.
+ * </p>
+ * <p>
+ *     The only hardship this imposes is if a confusingly-designed enum has values which differ only in case <em>and</em> the
+ *     string in the config source does not exactly match one of the enum value names. In such cases
+ *     the mapper will be unable to choose which enum value matches an ambiguous string. A developer faced with this
+ *     problem can simply provide her own explicit config mapping for that enum, for instance as a function parameter to
+ *     {@code Config.as}.
+ * </p>
+ *
+ */
+class EnumMapperProvider implements ConfigMapperProvider {
```

Comment:
You've raised two issues.

1. Why a map? For efficiency so we don't instantiate multiple identical function objects for the same `enum` type if user code requests conversion of the same `enum` more than once.

2. (by "a manager" do you mean "a mapper provider"?) 

If you know of a way to do this without a mapper provider I'm all ears.

From what I found in the config code (and the inner workings of config are rather new to me so I might have missed things), the single-mapper approach does not work for `enum` support. Each mapper works with a given type _all instances of which obey the same mapping rules_ (such as the mapper for `Integer`, the mapper for `Long`, etc.). Mappers are therefore passed only the `Config` node at conversation time...nothing about the specific type being converted to because the config infrastructure assumes the mapper knows which type it is dealing with and so knows what conversion rules to apply.

In contrast, with the `enum` case each `enum` is a _different_ Java type, not a different instance of the same Java type. So a single mapper for all `enum` types does not work because a single mapper is not given enough information at mapping time to know _which specific_ `enum` type a value needs to be mapped to.